### PR TITLE
Issue 41 -- Handle cmd arguments on lo2s -a 

### PR DESCRIPTION
**Pull request by cvonelm**
This pull requests adds the ability to measure in -a mode as long as a process, either given via the command line and executed by lo2s or as a foreign pid, is running. This goal was achieved by splitting up the ProcessMonitor class into its two components process monitoring and process control ( #41  has a more sincere explanation of what i've done).

### CODE REVIEW

File: src/monitor/process_monitor.cpp [hunk 1]
```diff
@@ -20,292 +20,75 @@
...
-        int status;
-        pid_t child = waitpid(-1, &status, __WALL);
-
-        num_wakeups_++;
+    Process &process = process_map().get_process(pid);
 
-        handle_signal(child, status);
+    if(process.info == nullptr)
+    {
+        process.info = new ProcessInfo(pid, false);
```
Comment (bmario) [inline src/monitor/process_monitor.cpp]: `new`

Use std::unique_ptr<...> instead and encapsulate this whole allocation logic inside the respective class.
This is called RAII and is the fundamental principle of resource management in C++

File: include/lo2s/monitor/dummy_monitor.hpp [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+#include <sys/types.h>
+}
+
+namespace lo2s
+{
+
+namespace monitor
+{
+
+class DummyMonitor
```
Comment (tilsche) [inline include/lo2s/monitor/dummy_monitor.hpp]: Please split this class in a pure virtual `AbstractProcessMonitor` and a dummy implementation `DummyProcessMonitor`. That way we won't get silent issues if we change/add methods to the base class of the actual `ProcessMonitor`.

**Comment by tilsche** [comment]
Looks good to me, like to wait for bmario's comment next week before the merge.

**Comment by cvonelm** [comment]
Sorry if I understood something wrong there, but currently the calls to make_unique are made inside the ProcessMonitor and not in the Process or Thread class, to ensure a clean separation between the bookkeeping part of the ProcessController and the monitoring part (this is why part of why I have made a Process and Thread class, which are currently somewhat redundant, as they only contain the pid or tid/ppid respectively). I know that this is somewhat hypothetical at the moment but if we would make a new Process Monitoring mode using the existing ThreadMonitor ecosystem (for example a SystemModeThreadMonitor) we would have to make either a new pair of Thread and Process classes which create SystemModeThreadMonitors or have 2 functions inside the Process class (start_process_monitor and start_system_monitor for example), which create the different monitors, which i personally like less than having the ProcessMonitor and SystemModeProcessMonitor create the respective objects for us.

**Comment by bmario** [comment]
Maybe I wasn't clear enough. Let's look for example at the class `Process`. While it is certainly a good idea to have the ProcessInfo member inside of an unique_ptr, I don't see a reason, why this class doesn't provide a member function like this:

```
void initialize_process_info(pid_t pid, bool spawn)
{
    if(!info)
       info = std::make_unique<ProcessInfo>(pid, spawn);
}
```

You're scattering class internals (how is the member initialized), which is hard to maintain.

**Comment by cvonelm** [comment]
Ok, that does make sense to me

**Comment by cvonelm** [comment]
I've came to the conclusion that ProcessMonitor and ProcessController share little to none information, so that the ProcessMap is quite superfluous, and replaced it with a simpler solution where ProcessMonitor and ProcessController do not share their Process and Thread bookkeeping.

**Pull request merged**